src/homos.c

```diff
@@ -1,1816 +1,1861 @@
+    FOR_SET_BITS(possible, GRAPH2->nr_vertices, i) {
+      MAP[next] = i;
+      set_bit_array(VALS, i, true);
+      set_bit_array(MAP_UNDEFINED[depth], next, false);
+      if (!has_trivial_stab) {
+        find_graph_homos(depth + 1,
+                         next,
+                         rep_depth + 1,
+                         compute_stabs_and_orbit_reps(GRAPH1->nr_vertices,
+                                                      GRAPH2->nr_vertices,
+                                                      rep_depth + 1,
+                                                      depth,
+                                                      i),
+                         rank + 1,
+                         max_results,
+                         hint,
+                         count);
+      } else {
+        find_graph_homos(depth + 1,
+                         next,
+                         rep_depth,
+                         true,
+                         rank + 1,
+                         max_results,
+                         hint,
+                         count);
+      }
+      MAP[next] = UNDEFINED;
+      set_bit_array(VALS, i, false);
+      set_bit_array(MAP_UNDEFINED[depth], next, true);
+    }
+    END_FOR_SET_BITS
+  }
+  copy_bit_array(
+      possible, get_conditions(CONDITIONS, next), GRAPH2->nr_vertices);
+  intersect_bit_arrays(possible, VALS, GRAPH2->nr_vertices);
+  FOR_SET_BITS(possible, GRAPH2->nr_vertices, i) {
+    MAP[next] = i;
+    set_bit_array(MAP_UNDEFINED[depth], next, false);
+    find_graph_homos(depth + 1,
+                     next,
+                     rep_depth,
+                     has_trivial_stab,
+                     rank,
+                     max_results,
+                     hint,
+                     count);
+    MAP[next] = UNDEFINED;
+    set_bit_array(MAP_UNDEFINED[depth], next, true);
+  }
+  END_FOR_SET_BITS
+  pop_conditions(CONDITIONS, depth);
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-// is_adjacent_graph: returns <true> if there is an edge from <i> to <j> in the
-// Graph pointed to by <graph> and returns <false> if there is not.
+// The main recursive function for injective homomorphisms of graphs
 ////////////////////////////////////////////////////////////////////////////////
 
-inline static bool is_adjacent_graph(Graph* graph, Vertex i, Vertex j) {
-  DIGRAPHS_ASSERT(graph != NULL);
-  DIGRAPHS_ASSERT(i < graph->nr_vertices && j < graph->nr_vertices);
-
-  return get_bit_array(graph->neighbours[i], j);
+static ALWAYS_INLINE uint16_t
+                     graph_mono_update_conditions(uint16_t const depth,
+                                                  uint16_t const last_defined,
+                                                  uint16_t const vertex) {
+  push_conditions(
+      CONDITIONS, depth, vertex, GRAPH2->neighbours[MAP[last_defined]]);
+  set_bit_array(get_conditions(CONDITIONS, vertex), MAP[last_
```

Comment:
digraph2_obj -> digraph1_obj

Reply:
The preview here seems to be completely the wrong line, my comment is for line 1725